src/lib.rs

```diff
@@ -150,31 +169,125 @@ fn handle_block(
         check_receipt_root(&block)?;
         check_transaction_root(&block)?;
     }
-    
+    //TODO: Fix outputing
+    // if let Some(output) = output {
+    //     let file_name = format!("{}/block-{}.json", output, block.number);
+    //     let mut out_file = File::create(file_name)?;
 
-    if let Some(output) = output {
-        let file_name = format!("{}/block-{}.json", output, block.number);
-        let mut out_file = File::create(file_name)?;
+    //     let block_json = protobuf_json_mapping::print_to_string(&block)
+    //         .map_err(|err| DecodeError::ProtobufError(err.to_string()))?;
+
+    //     out_file
+    //         .write_all(block_json.as_bytes())
+    //         .map_err(DecodeError::IoError)?;
+    // }
 
-        let block_json = protobuf_json_mapping::print_to_string(&block)
-            .map_err(|err| DecodeError::ProtobufError(err.to_string()))?;
+    Ok(block)
+}
+
+pub fn extract_blocks<R: Read>(mut reader: R) -> Result<Vec<Block>, DecodeError> {
+    log::debug!("Reading messages");
+    let dbin_file = DbinFile::try_from_read(&mut reader)?;
+    log::debug!("Validating blocks");
+
+    // Parallel processing of block headers
+    dbin_file
+        .messages
+        .par_iter()
+        .map(|message| handle_block(message, None, None))
+        .collect()
+}
 
-        out_file
-            .write_all(block_json.as_bytes())
-            .map_err(DecodeError::IoError)?;
+// pub fn stream_blocks<R: Read, W: Write>()
+// A function which decodes blocks from a reader and writes them, serialized, to a writer
+pub async fn stream_blocks<R: Read, W: Write>(
+    mut reader: R,
+    mut writer: W,
+    end_block: Option<usize>,
+) -> Result<(), DecodeError> {
+    let end_block = match end_block {
+        Some(end_block) => end_block,
+        None => MERGE_BLOCK,
+    };
+    let mut block_number = 0;
+    loop {
+        match DbinFile::read_message_stream(&mut reader) {
+            Ok(message) => {
+                let block = decode_block_from_bytes(&message)?;
+                block_number = block.number as usize;
+
+                if block_number != 0 {
```

Comment:
Yeah, at one point, there was a problem decoding the first block. It looks like this has since been fixed, probably when you, @srosati, added support for pre-byzantium receipts. I've removed this skip in a8311a8d5d481aa25787e88f90869d83c889d367.